dao/courses.go

```diff
@@ -281,6 +282,32 @@ func (d coursesDao) GetCourseAdmins(courseID uint) ([]model.User, error) {
 	return admins, err
 }
 
+type Course struct {
+	Name, Slug, TeachingTerm, Visibility string
+	ID                                   uint
+	Year                                 int
+}
+
+// ExecAllCourses executes f on all courses.
+func (d coursesDao) ExecAllCourses(f func([]Course)) {
+	var res []Course
+	batchNum := 0
+	batchSize := 100
+	var numCourses int64
+	DB.Model(&model.Course{}).Count(&numCourses)
+	for batchSize*batchNum < int(numCourses) {
+		err := DB.Raw(`SELECT id, name, slug, year, teaching_term, visibility
+							FROM courses 
+							WHERE deleted_at IS NULL
+							LIMIT ? OFFSET ?`, batchSize, batchNum*batchSize).Scan(&res).Error
```

Comment:
Please add a `ORDER BY ID` to avoid any undefined behaviour with limit/offset based pagination.